AuxiliaryConnections to Nodules

mattigruener opened the pull request:
Hey John,

I took a stab at implementing what we had discussed about the `plugWithoutNodule->plugWithNodule` case. I think what I have now is very close to your suggestion. You had drawn an image using straight lines and instead I'm abusing @danieldresser's noodle rendering code to render the connection more similarly to what our other noodles look like. Integrates a little better with the look of the node gadgets and the other noodles?

![Foo](https://github.com/GafferHQ/gaffer/blob/b242a9e08bd04e04c6de97df9cf768399b461c89/auxConnection.png?raw=true)

I am unsure if you like what I've done in the first half of the commits here. Messing with the shader code was much easier when I was able to reload it. I've committed the changes I've made in that regard so that the next person's life gets a little easier. Are you opposed to 
- keeping the shaders in separate files and 
- exposing a way to trigger shader reloading?

Moving shaders and then modifying them is in separate commits to make reviewing easier.

The rest is hopefully fairly close? It's a little messy to convert everything from *3f to *2f and back only to pass the data through the Style API. We're moving towards a 2D API though, right? It's correct to have the Style accept 2D data even if that's creating some overhead in the caller and the callee?

Cheerio! :)

Matti.

danieldresser-ie commented:
There was a slight miscommunication between Matti and I - I was picturing a slightly different implementation : rather than adding more complexity to the already rather complicated noodle drawing code, I was picturing making two GL calls - one standard noodle with the endPoint offset from outside, and then a separate call to draw the extra little extension straight segment.  Not a big deal to me, but it seems like it might look a little bit cleaner.

Though, actually, looking at it now, why does the initial segment even need to be straight?  If we just used the same turning radius as all the other noodles, it seems like it would look fine?

johnhaddon commented on glsl/Gadgets.vert:
```diff
@@ -0,0 +1,57 @@
+  gl_FrontColor = gl_Color;
+  gl_BackColor = gl_Color;
+  gl_TexCoord[0] = gl_MultiTexCoord0;
+};
```
Your nemesis (the GL drivers on my Mac) say :

```
RuntimeError: Exception : ERROR: 0:82: ';' : syntax error: syntax error
```

Same for the .frag file.

johnhaddon commented on glsl/Gadgets.vert:
```diff
@@ -0,0 +1,57 @@
+
+void main()
+{
+  if( isCurve )
```
Tabs not spaces.

johnhaddon commented on src/GafferUI/StandardConnectionGadget.cpp:
```diff
@@ -186,6 +186,35 @@ void StandardConnectionGadget::doRenderLayer( Layer layer, const Style *style )
+				const Box3f b = srcNodeGadget->transformedBound();
+				const Box2f srcNodeFrame( V2f( b.min.x, b.min.y ), V2f( b.max.x, b.max.y ) );
+
+				const M44f m = dstNodule()->fullTransform( srcNodeGadget );
```
Do we really need to duplicate the logic from `setPositionsFromNodules()` here? What's wrong with `m_dstPos` and `m_dstTangent`?

johnhaddon commented on src/GafferUI/StandardConnectionGadget.cpp:
```diff
@@ -186,6 +186,35 @@ void StandardConnectionGadget::doRenderLayer( Layer layer, const Style *style )
+				const V3f t = dstNodeGadget->noduleTangent( dstNodule() );
+				const V2f dstTangent = V2f( t.x, t.y );
+
+				Style::State state = srcNodeGadget->getHighlighted() || dstNodeGadget->getHighlighted() ? Style::HighlightedState : Style::NormalState;
```
Again, why duplicate logic that already exists below (line 220 onwards)?

There's a related question here I think. Currently it's possible to start a drag with one of these auxiliary connections (only from the dst end as far as I can tell), but there's no visual indication that it's happening at all, until you drop into empty space and the connection is deleted. Do we want to allow that? To me it seems like it could be potentially useful, and I think providing the visual feedback is simply a case of removing the duplicate logic from in here.

I suspect the Expression node might not be super happy if the drag allowed an output to be connected into two places, but that will be fixed by the upcoming "OSLCode-stylee" expressions we've been talking about.

johnhaddon commented on glsl/Gadgets.frag:
```diff
@@ -0,0 +1,61 @@
+
+#endif
+
+void main()
```
This shader now appears in the Scene/OpenGL/Shaders submenu! I don't think we want that. Also, I think StandardStyle.frag might be a more descriptive name...

johnhaddon commented on include/GafferUI/StandardStyle.h:
```diff
@@ -98,6 +98,8 @@ class GAFFERUI_API StandardStyle : public Style
 		void renderRotateHandle( Axes axes, State state = NormalState ) const override;
 		void renderScaleHandle( Axes axes, State state = NormalState ) const override;
 
+    static void reloadShaders();
```
Indentation. I know it's fixed in a later commit, but it still makes review harder...

mattigruener commented:
Hey John,

I've removed all shader shenanigans, dot-adding is disabled and dragging these `AuxiliaryConnections` should work now. I addressed your comments regarding code duplications, too. Is this ok, or do you see anything that can be improved still?

Before merging this, I was curious to hear your opinion on changing the `StandardStyle` API one last time... We had agreed on the `(Box2f, V2f, V2f)` signature for this new method, but with these connections behaving more like "normal" StandardConnections now, maybe it's worth thinking about matching the signature of `renderConnection()`? I'm mostly bringing it up because I'm anticipating a fix in `AuxiliaryConnectionsGadget` for connections that are the other way around - connections whose src has a nodule and the dst doesn't. Currently we render those as straight `AuxiliaryConnections` between the `NodeGadgets`, which seems fine for now, but I'm assuming we'll want to address that eventually. I thought I'd ask before adding a commit that makes that change.

Cheerio,

Matti.

johnhaddon commented:
> Before merging this, I was curious to hear your opinion on changing the StandardStyle API one last time... We had agreed on the (Box2f, V2f, V2f) signature for this new method, but with these connections behaving more like "normal" StandardConnections now, maybe it's worth thinking about matching the signature of renderConnection()?

Good idea - let's do that.

mattigruener commented:
> Good idea - let's do that.

Ok great, how's the latest code?

It feels a little sketchy to me to rely on [the setup for stubs](https://github.com/GafferHQ/gaffer/blob/master/src/GafferUI/StandardConnectionGadget.cpp#L148) to provide a proper tangent for the rendering of AuxiliaryConnections, but maybe that's alright..?

Thanks :)

Matti.

johnhaddon commented on src/GafferUI/StandardConnectionGadget.cpp:
```diff
@@ -188,24 +188,46 @@ void StandardConnectionGadget::doRenderLayer( Layer layer, const Style *style )
+	{
+		Box3f b = srcNodeGadget->transformedBound();
+		style->renderAuxiliaryConnection( V2f( b.center().x, b.center().y ), V2f( m_srcTangent.x, m_srcTangent.y ), V2f( m_dstPos.x, m_dstPos.y ), V2f( m_dstTangent.x, m_dstTangent.y ), state );
+		return;
```
This early return skips the drawing of the extra connection when `m_addingConnection == true`. Actually, I'm a bit confused, but I think your branch is from a point before we merged the `m_addingConnection` stuff?

johnhaddon commented on src/GafferUI/StandardConnectionGadget.cpp:
```diff
@@ -478,7 +504,7 @@ void StandardConnectionGadget::updateDotPreviewLocation( const ButtonEvent &even
 bool StandardConnectionGadget::keyPressed( const KeyEvent &event )
 {
-	if( event.modifiers & ButtonEvent::Control )
+	if( event.modifiers & ButtonEvent::Control && srcNodule() && dstNodule() )
```
Note that `dstNodule()` is actually [guaranteed to be non-null](https://github.com/GafferHQ/gaffer/blob/master/src/GafferUI/ConnectionGadget.cpp#L92). I think you've picked up this "belt and braces" approach from code I wrote earlier in the file, but it's better to omit unnecessary guards because they spread confusion and mask bugs.

johnhaddon commented:
Closed, since we've gone with the variation in #2497 instead...
